### PR TITLE
New lesson form

### PR DESCRIPTION
**Pull request by chinwan34**
I have implemented the following new functionality for the lesson request page, the students are able to delete and update the lesson requests, while the admin is able to approve or deny the lesson requests.

The follow pages are created:
- lesson_request_update.py
- lesson_request_delete.py

### CODE REVIEW

File: lessons/helpers/helper_functions.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ def get_user_group(request):
         raise Exception("User groups do not exist. Try running 'python3 manage.py create_user_groups'")
 
     for key in USER_GROUPS.keys():
+        print(key+"\n")
```
Comment (Ali-Alkhars) [inline lessons/helpers/helper_functions.py]: This is why the tests print the user type. You should delete it

File: lessons/models.py [hunk 2]
```diff
@@ -20,6 +21,59 @@ class User(AbstractUser):
...
+            ('%(value) is not allowed'),
+            params={'value':value},
+        )
+
+
+class Lesson_request(models.Model):
+    student = models.ForeignKey(User, on_delete=models.CASCADE, null=True, blank=True)
+    availableDays = models.CharField(max_length=100, blank=False, choices=weekday_choices)
+    availableTimes = models.CharField(max_length=255, blank=False, choices=time_choices)
+    # availableDays = models.ManyToManyField(Weekdays_Choices, symmetrical=False)
```
Comment (Ali-Alkhars) [inline lessons/models.py]: Remove commented out code

File: lessons/models.py [hunk 3]
```diff
@@ -20,6 +21,59 @@ class User(AbstractUser):
...
+
+fulfilled_choices = [
+    ('Approved', 'Approved'),
+    ('Denied', 'Denied'),
+    ('Pending', 'Pending'),
+]
+
+value_choices = [
+]
+for i in range(1, 17):
```
Comment (Ali-Alkhars) [inline lessons/models.py]: I don't think it's clean code to put a for loop out here like this haha! You could just write down the choices (up to 120 minutes) yourself to make it cleaner

File: lessons/models.py [hunk 4]
```diff
@@ -20,6 +21,59 @@ class User(AbstractUser):
...
+    LearningObjectives = models.TextField(blank=False)
+    AdditionalNotes = models.TextField(blank=True)
+    Fulfilled = models.CharField(max_length=50, blank=False, choices=fulfilled_choices, default='Pending')
+
+    def getWeekdays_choices():
+        return weekday_choices
+    
+    def getTime_choices():
+        return time_choices
     def isStudent(self):
```
Comment (Ali-Alkhars) [inline lessons/models.py]: The isStudent, isAdmin and isDirector methods were in the User model class. You might have moved them by mistake, so please make sure to put them back in the User class

File: lessons/templates/lesson_page.html [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                {% if currentUser == False %}
+                <td><a href="{% url 'lesson_request_update' user.id %}">Edit</a></td>
+                <td><a href="{% url 'lesson_request_delete' user.id%}">Delete</a></td>
+                {% endif %}
+                
+            </tr>
+            {% endfor %}
+        </table>
+            
+        <p><a href='{%url 'lesson_request' %}' class="btn btn-lg btn-secondary"><button>Request a new lesson</button></a></p>
```
Comment (Ali-Alkhars) [inline lessons/templates/lesson_page.html:57]: This button should only show for students. You could add the following:
```
{% if user.isStudent %}
      The button is added here
{% endif %}
```

All templates already have the user object so you don't need to pass it in the view

File: lessons/models.py [hunk 6]
```diff
@@ -20,6 +21,59 @@ class User(AbstractUser):
     first_name = models.CharField(max_length=20, blank=False)
     last_name = models.CharField(max_length=20, blank=False)
 
+
+weekday_choices = [
```
Comment (Ali-Alkhars) [inline lessons/models.py]: All constant fields should be in ALL CAPS, it's a python convention

File: lessons/views.py [hunk 7]
```diff
@@ -1,14 +1,90 @@
...
+            return redirect("lesson_page")
+    else:
+        form = LessonRequestForm()
+    return render(request, "lesson_request.html", {"form":form})
+
+"""
+Check all the requests and bookings of lessons
+"""
+@login_required
+@permitted_groups(['student', 'admin', 'director'])
```
Comment (Ali-Alkhars) [inline lessons/views.py]: You don't need to add a decorator if all user groups can access the page

File: lessons/views.py [hunk 8]
```diff
@@ -1,14 +1,90 @@
...
+        count = Lesson_request.objects.all().count()
+        request.session['countOfTable'] = count
+        data = {'object_list': Lesson_request.objects.all(), 'count': count, 'currentUser': isStudent}
+    return render(request, "lesson_page.html", data)
+
+"""
+Update a particular lesson request
+"""
+@login_required
+@permitted_groups(['student', 'admin', 'director'])
```
Comment (Ali-Alkhars) [inline lessons/views.py]: You don't need to add a decorator if all user groups can access the page

File: lessons/forms.py [hunk 9]
```diff
@@ -1,9 +1,30 @@
...
 from django.core.validators import RegexValidator
-from lessons.models import User
+from lessons.models import User, Lesson_request
 from django.contrib.auth.models import Group
 
+# Request form for the lesson
+class LessonRequestForm(forms.ModelForm):
+    class Meta:
+        model = Lesson_request
+        fields = ["availableDays", "availableTimes", "numberOfLessons", "IntervalBetweenLessons", "DurationOfLesson", "LearningObjectives", "AdditionalNotes", "Fulfilled"]
```
Comment (Ali-Alkhars) [inline lessons/forms.py]: I don't think 'Fulfilled' should be in the form

**Review by Ali-Alkhars — CHANGES_REQUESTED** [review]
Well done overall! I think this does the job but needs some changes which I pointed out in several comments, so please check those out first.

**Here are two additional things:**
- You should add a link to the 'lesson_page' to the navbar partial template
- For admins and directors. To approve/deny a lesson request, they shouldn't have to go to the edit page then change the fulfilled value (Fulfilled shouldn't even be in the form, I mentioned it in one of the comments). So instead, there should be two buttons that appear only for the admin/director. Approve button and Deny button, along with the edit and delete

File: lessons/helpers/helper_functions.py (hunk 1 above)
Comment (chinwan34) [reply lessons/helpers/helper_functions.py]: Successfully deleted.

File: lessons/models.py (hunk 2 above)
Comment (chinwan34) [reply lessons/models.py]: Successfully commented out.

File: lessons/models.py (hunk 3 above)
Comment (chinwan34) [reply lessons/models.py]: Successfully changed to manual type-in.

File: lessons/models.py (hunk 4 above)
Comment (chinwan34) [reply lessons/models.py]: Successfully repositioned it.

File: lessons/models.py (hunk 6 above)
Comment (chinwan34) [reply lessons/models.py]: All CAPS used.

File: lessons/views.py (hunk 7 above)
Comment (chinwan34) [reply lessons/views.py]: Deleted unnecessary decorator.

File: lessons/views.py (hunk 8 above)
Comment (chinwan34) [reply lessons/views.py]: Deleted unnecessary decorator.

File: lessons/forms.py (hunk 9 above)
Comment (chinwan34) [reply lessons/forms.py]: I left Fulfilled in the model, but make it invisible in the form just like the student field. Then I implemented the approve / deny functions that is only visible by admin/director.

File: lessons/templates/lesson_page.html [hunk 10]
```diff
@@ -23,32 +23,34 @@
...
+                {% if user.isStudent %}
+                {% if form.Fulfilled == "Pending" %}
+                <td><a href="{% url 'lesson_request_update' form.id %}">Edit</a></td>
+                <td><a href="{% url 'lesson_request_delete' form.id%}">Delete</a></td>
                 {% endif %}
                 {% endif %}
-                {% if currentUser == False %}
-                <td><a href="{% url 'lesson_request_update' user.id %}">Edit</a></td>
-                <td><a href="{% url 'lesson_request_delete' user.id%}">Delete</a></td>
+                {% if user.isStudent == False %}
```
Comment (Ali-Alkhars) [inline lessons/templates/lesson_page.html:43]: Here I think you might have misunderstood what I meant. Even though you made the approve/deny buttons like I suggested, I didn't mean that you should delete the edit/delete buttons for the admin or director.

So they should always be able to edit/delete a request, however, the approve/deny buttons should not appear after the lesson is approved or denied. I think just a simple if statement to check if the lesson is pending should do it.

**Review by Ali-Alkhars — COMMENTED** [review]
Well done! Just one small little tiny thing needs changing though, check the comment

**Review by Ali-Alkhars — APPROVED** [review]
Well done!

**Pull request merged**